File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
#include <bits/stdc++.h>
using namespace std;


class Solution {
public:

    int lengthOfLongestSubstring(string s) {
     // variable size sliding window
     // use a hashset to see whether element is in range
     int n = s.size();
     unordered_set<char> chars;
     int l = 0;
     int ans = 0;
     for (int r = 0; r < n; ++r) {
         // if not found just add it and maximize
         if (!chars.count(s[r])) {
             chars.insert(s[r]);
             ans = max(ans, r - l + 1);
         }
         else {
             while (chars.count(s[r])) {
                 chars.erase(s[l]);
                 ++l;
             }
             chars.insert(s[r]);
         }
     }
     
     return ans;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let ans = 0, cur = 0, i = 0, n = s.length;
    let seen = [];

    while (i < n) {
        if (!seen.includes(s[i]))
            seen.push(s[i]), cur++;
        else {
            while(seen[0] !== s[i])
                seen.shift(), --cur;
            seen.shift(), --cur;
            seen.push(s[i]), ++cur;
        }
        ++i;
        ans = Math.max(ans, cur);
    }
    
    return ans;
};
